Created new section called Parallel batch execution

nataliefiann opened the pull request:
## What are you changing in this pull request and why?

I have created this PR following this Git issue: https://github.com/dbt-labs/docs.getdbt.com/issues/6550 for Parallel batch execution

## Checklist
- [ ] I have reviewed the [Content style guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) so my content adheres to these guidelines.
- [ ] The topic I'm writing about is for specific dbt version(s) and I have versioned it according to the [version a whole page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version) and/or [version a block of content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content) guidelines.
- [ ] I have added checklist item(s) to this list for anything anything that needs to happen before this PR is merged, such as "needs technical review" or "change base branch."
- [ ] The content in this PR requires a dbt release note, so I added one to the [release notes page](https://docs.getdbt.com/docs/dbt-versions/dbt-cloud-release-notes).
<!--
PRE-RELEASE VERSION OF dbt (if so, uncomment):
- [ ] Add a note to the prerelease version [Migration Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/docs/dbt-versions/core-upgrade)
-->
<!-- 
ADDING OR REMOVING PAGES (if so, uncomment):
- [ ] Add/remove page in `website/sidebars.js`
- [ ] Provide a unique filename for new pages
- [ ] Add an entry for deleted pages in `website/vercel.json`
- [ ] Run link testing locally with `npm run build` to update the links that point to deleted pages
-->

<!-- vercel-deployment-preview -->
---
🚀 Deployment available! Here are the direct links to the updated files:


- https://docs-getdbt-com-git-nfiann-rbip-dbt-labs.vercel.app/docs/build/incremental-microbatch
- https://docs-getdbt-com-git-nfiann-rbip-dbt-labs.vercel.app/docs/dbt-versions/core-upgrade/06-upgrading-to-v1.9
- https://docs-getdbt-com-git-nfiann-rbip-dbt-labs.vercel.app/reference/resource-properties/concurrent_batches

<!-- end-vercel-deployment-preview -->

nataliefiann committed:
Created new section called Parallel batch execution

mirnawong1 committed:
Merge branch 'current' into nfiann-rbip

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
````diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+select ...
+```
+
+</File>
````
Note - I need to double check this sql example.

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -185,6 +185,7 @@ Several configurations are relevant to microbatch models, and some are required:
 | `begin`      | Date (required)   | The "beginning of time" for the microbatch model. This is the starting point for any initial or full-refresh builds. For example, a daily-grain microbatch model run on `2024-10-01` with `begin = '2023-10-01` will process 366 batches (it's a leap year!) plus the batch for "today."        | N/A     |
 | `batch_size` | String (required)  | The granularity of your batches. Supported values are `hour`, `day`, `month`, and `year`             | N/A     |
 | `lookback`   | Integer (optional) | Process X batches prior to the latest bookmark to capture late-arriving records.                                         | `1`     |
+|[`concurrent_batches`](docs/build/incremental-microbatch#concurrent-execution-with-the-concurrent_batches-configuration)|Boolean|Determines whether batches should be run concurrently (at the same time) or sequentially (one after the other)|`True` or `False`|
```
```suggestion
|`concurrent_batches`| Boolean | Configures whether batches can run concurrently (at the same time) or sequentially (one after the other). When set to `True` and conditions are met, batches run in parallel. |`True` or `False`|
```

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
 
+## Parallel batch execution
+
+The microbatch strategy offers the benefit of updating a model in smaller, more manageable batches. Within dbt, these microbatches can often run at the same time (in parallel), enhancing efficiency.
```
the last line is similar to the first line in the 2nd paragraph so sugg removing maybe
```suggestion
The microbatch strategy offers the benefit of updating a model in smaller, more manageable batches. 
```

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+
+The microbatch strategy offers the benefit of updating a model in smaller, more manageable batches. Within dbt, these microbatches can often run at the same time (in parallel), enhancing efficiency.
+
+Parallel batch execution means that multiple batches are processed at the same time, instead of one after the other (sequentially). For example, if you have a micro-batch model with 12 batches, those batches can be executed in parallel. Specifically they'll run in parallel limited by the number of [available threads](docs/running-a-dbt-project/using-threads).
```
active voice and broken link
```suggestion
Parallel batch execution means that multiple batches are processed at the same time, instead of one after the other (sequentially). For example, if you have a microbatch model with 12 batches, you can execute those batches in parallel. Specifically they'll run in parallel limited by the number of [available threads](/docs/running-a-dbt-project/using-threads).
```

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+
+Parallel batch execution means that multiple batches are processed at the same time, instead of one after the other (sequentially). For example, if you have a micro-batch model with 12 batches, those batches can be executed in parallel. Specifically they'll run in parallel limited by the number of [available threads](docs/running-a-dbt-project/using-threads).
+
+To run batches in parallel you can use the `concurrent_batches` configuation:
```
```suggestion
To run batches in parallel, use the `concurrent_batches` configuration:
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+{{
+  config(
+    materialized='incremental',
+    incremental_strategy='concurrent_batches'
```
i think you should double check this - the config seems wrong and the strategy isn't concurrent_batches right? it should be microbatch? 

see: https://arc.net/l/quote/maxltpur

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+
+It's not possible to run batches concurrently when the model definition contains [`{{ this }}`](/reference/dbt-jinja-functions/this). This is because Jinja represents the state of a table.
+
+### Concurrent Execution with the `concurrent_batches` configuration
```
```suggestion
### Configure `concurrent_batches`
```

mirnawong1 replied:
what's the difference between this header ( Concurrent Execution with the `concurrent_batches` configuration) and the config example mentioned earlier? 

maybe this is more order of operations or  parallel execution sequence?

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+| `False`                    |Batches will run sequentially.|
+| `None` (not explicitly set)|dbt will evaluate the model's Jinja logic and evaluate if it contains a reference to `this`. If it references `this`, the batches will run sequentially. This is because Jinja represents states of table. Otherwise, the batches will run in parallel.|
+
+### Sequential Execution Conditions
```
against style guide - we use sentence case
```suggestion
### Sequential execution conditions
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+1. The database adapter does not support concurrent batches.
+2. The relation in the data warehouse for the model doesn't exist.
+
+### Supported adapters
```
do we think this should be higher up and mentioned at the beginning so users know this ahead of time?

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+
+### Concurrent Execution with the `concurrent_batches` configuration
+
+To enable parallel execution when possible, users can check the model's concurrent_batches configuration:
```
```suggestion
To enable parallel execution when possible, users can check the model's `concurrent_batches` configuration:
```

mirnawong1 commented:
hey @nataliefiann , good job! it's getting there -- remember you still want to add this to the:

- 1.9 upgrade guide
- add a mention about parallel execution [here](https://docs.getdbt.com/docs/build/incremental-microbatch#what-is-microbatch-in-dbt) and link to the section
- @QMalcolm @nataliefiann since we're adding new pages for the specific configs (`lookback` and `batch_size`) , think `concurrent_batches` should prob have its own config page too, rigth?

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
````diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+
+```yaml
+models:
+  +incremental_strategy: "concurrent_batches"
````
i dont think this is right

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
````diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+```
+
+</File>
+
````
```suggestion
If any of the following conditions are met, the batches will execute sequentially:

1. The database adapter does not support concurrent batches.
2. The relation in the data warehouse for the model doesn't exist.
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,71 @@ For now, dbt assumes that all values supplied are in UTC:
+
+### Sequential execution Conditions
+
+If any of the following conditions are met, the batches will execute sequentially:
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+
+</File>
+
+It's not possible to run batches concurrently when the model definition contains [`{{ this }}`](/reference/dbt-jinja-functions/this). This is because Jinja represents the state of a table.
```
do you think the `{{this}}` should be added in the table too so its not lost? just thinking whether all the caveats need to be in one place so users have the required info in one space/section/place.

mirnawong1 replied:
![Screenshot 2024-12-05 at 09 37 49](https://github.com/user-attachments/assets/929c64ef-4b39-48f9-8215-38e07bd696df)
this table here

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,70 @@ For now, dbt assumes that all values supplied are in UTC:
+### Sequential execution Conditions
+
+
+1. The database adapter does not support concurrent batches.
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,69 @@ For now, dbt assumes that all values supplied are in UTC:
+### Sequential execution Conditions
+
+
+2. The relation in the data warehouse for the model doesn't exist.
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +240,68 @@ For now, dbt assumes that all values supplied are in UTC:
+| `False`                    |Batches will run sequentially.|
+| `None` (not explicitly set)|dbt will evaluate the model's Jinja logic and evaluate if it contains a reference to `this`. If it references `this`, the batches will run sequentially. This is because Jinja represents states of table. Otherwise, the batches will run in parallel.|
+
+### Sequential execution Conditions
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
````diff
@@ -239,6 +240,67 @@ For now, dbt assumes that all values supplied are in UTC:
+```
+
+</File>
+If any of the following conditions are met, the batches will execute sequentially:
````
```suggestion
### Sequential execution Conditions

If any of the following conditions are met, the batches will execute sequentially:
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -185,6 +185,7 @@ Several configurations are relevant to microbatch models, and some are required:
 | `begin`      | Date (required)   | The "beginning of time" for the microbatch model. This is the starting point for any initial or full-refresh builds. For example, a daily-grain microbatch model run on `2024-10-01` with `begin = '2023-10-01` will process 366 batches (it's a leap year!) plus the batch for "today."        | N/A     |
 | `batch_size` | String (required)  | The granularity of your batches. Supported values are `hour`, `day`, `month`, and `year`             | N/A     |
 | `lookback`   | Integer (optional) | Process X batches prior to the latest bookmark to capture late-arriving records.                                         | `1`     |
+|`concurrent_batches`| Boolean | Configures whether batches can run concurrently (at the same time) or sequentially (one after the other). When set to `True` and conditions are met, batches run in parallel. |`True` or `False`|
```
```suggestion
|`concurrent_batches`| Boolean | Configures whether batches can run concurrently (at the same time) or sequentially (one after the other). When set to `True` and conditions are met, batches run in parallel. |`True` or `False`|
| `{{this}}` | Macro | When the model definition contains `{{ this }}`, batches can't be ran concurrently | Refer to [Concurrent Execution with the `concurrent_batches` configuration](/docs/build/incremental-microbatch#concurrent-execution-with-the-concurrent_batches-configuration) for more information |
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +241,69 @@ For now, dbt assumes that all values supplied are in UTC:
+{{
+  config(
+    materialized='incremental',
+    incremental_strategy='concurrent_batches'
```
```suggestion
    incremental_strategy='microbatch'
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
````diff
@@ -239,6 +241,69 @@ For now, dbt assumes that all values supplied are in UTC:
+
+```yaml
+models:
+  +incremental_strategy: "concurrent_batches"
````
```suggestion
  +incremental_strategy: "microbatch"
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +241,69 @@ For now, dbt assumes that all values supplied are in UTC:
+{{
+  config(
+    materialized='incremental',
+    incremental_strategy='microbatch'
```
```suggestion
    unique_key='concurrent_batches',
    incremental_strategy='microbatch'
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented:
Hiya @QMalcolm 

I'm not 100% sure the config examples I left are correct. Would you mind reviewing them as well when you get the chance please?

Kind Regards
Natalie

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -185,6 +185,8 @@ Several configurations are relevant to microbatch models, and some are required:
 | `batch_size` | String (required)  | The granularity of your batches. Supported values are `hour`, `day`, `month`, and `year`             | N/A     |
 | `lookback`   | Integer (optional) | Process X batches prior to the latest bookmark to capture late-arriving records.                                         | `1`     |
+|`concurrent_batches`| Boolean | Configures whether batches can run concurrently (at the same time) or sequentially (one after the other). When set to `True` and conditions are met, batches run in parallel. |`True` or `False`|
+| `{{this}}` | Macro | When the model definition contains `{{ this }}`, batches can't be ran concurrently | Refer to [Concurrent Execution with the `concurrent_batches` configuration](/docs/build/incremental-microbatch#concurrent-execution-with-the-concurrent_batches-configuration) for more information |
```
Hmmmm, I'm not sure if we should include `{{ this }}` in the config table. It's not a config, but a [dbt jinja function](https://docs.getdbt.com/reference/dbt-jinja-functions/this). Maybe we can instead just cover this topic in parallel batch execution section?

mirnawong1 replied:
➕

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +241,70 @@ For now, dbt assumes that all values supplied are in UTC:
+The microbatch strategy offers the benefit of updating a model in smaller, more manageable batches. 
+
+Parallel batch execution means that multiple batches are processed at the same time, instead of one after the other (sequentially). For example, if you have a microbatch model with 12 batches, you can execute those batches in parallel. Specifically they'll run in parallel limited by the number of [available threads](/docs/running-a-dbt-project/using-threads).
+
```
```suggestion
### How it's determined if a batch will run in parallel

A batch can only run in parallel if:
1. It's **not** the first batch
2. It's **not** the last batch
3. The [adapter supports](/docs/build/incremental-microbatch#supported-adapters) concurrent batches

After [1], [2], and [3] we check if the [`this` jinja function](https://docs.getdbt.com/reference/dbt-jinja-functions/this) is invoked in the model. If `this` is used, then the batch will be run sequentially, as it may be that your batch depends on the existence of prior batches. If `this` isn't used, the batch will be run in parallel.

You can override the check for `this` by setting `concurrent_batches` to either `True` or `False`. If set to `False`, the batch will be run sequentially. If set to `True` the batch will be run in parallel (assuming [1], [2], and [3])
```

QMalcolm replied:
Also, perhaps we should link `The adapter supports concurrent batches` to the `### Supported adapters` section

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +241,70 @@ For now, dbt assumes that all values supplied are in UTC:
+
+Parallel batch execution means that multiple batches are processed at the same time, instead of one after the other (sequentially). For example, if you have a microbatch model with 12 batches, you can execute those batches in parallel. Specifically they'll run in parallel limited by the number of [available threads](/docs/running-a-dbt-project/using-threads).
+
+To run batches in parallel, use the `concurrent_batches` configuration:
```
```suggestion
To override the `this` check, use the `concurrent_batches` configuration:
```

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
````diff
@@ -239,6 +241,70 @@ For now, dbt assumes that all values supplied are in UTC:
+
+```yaml
+models:
+  +incremental_strategy: "microbatch"
````
```suggestion
  +concurrent_batches: True
```

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +241,70 @@ For now, dbt assumes that all values supplied are in UTC:
+{{
+  config(
+    materialized='incremental',
+    unique_key='concurrent_batches',
```
```suggestion
    concurrent_batches=True,
```

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -239,6 +241,70 @@ For now, dbt assumes that all values supplied are in UTC:
+2. The relation in the data warehouse for the model doesn't exist.
+It's not possible to run batches concurrently when the model definition contains [`{{ this }}`](/reference/dbt-jinja-functions/this). This is because Jinja represents the state of a table.
+
+### Configure `concurrent_batches`
+
+To enable parallel execution when possible, users can check the model's `concurrent_batches` configuration:
+
+| `concurrent_batches`       | Description                | 
+|----------------------------|----------------------------|
+| `True`                     |Batches will run in parallel.|
+| `False`                    |Batches will run sequentially.|
+| `None` (not explicitly set)|dbt will evaluate the model's Jinja logic and evaluate if it contains a reference to `this`. If it references `this`, the batches will run sequentially. This is because Jinja represents states of table. Otherwise, the batches will run in parallel.|
```
```suggestion
| `None` (default)|dbt will evaluate the model's Jinja logic and evaluate if it contains a reference to `this`. If it references `this`, the batches will run sequentially. This is because Jinja represents states of table. Otherwise, the batches will run in parallel.|
```

QMalcolm replied:
Should this be moved up to before the examples?

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
````diff
@@ -239,6 +241,70 @@ For now, dbt assumes that all values supplied are in UTC:
+```
+
+</File>
+### Sequential execution Conditions
+
+If any of the following conditions are met, the batches will execute sequentially:
+
+1. The database adapter does not support concurrent batches.
+2. The relation in the data warehouse for the model doesn't exist.
+It's not possible to run batches concurrently when the model definition contains [`{{ this }}`](/reference/dbt-jinja-functions/this). This is because Jinja represents the state of a table.
````
Replaced by `### How it's determined if a batch will run in parallel` section

QMalcolm reviewed (commented):
Thank you for doing this work ❤️ 

I don't claim to be a documentation writer, but It might be nice to move some things around 🙂

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -185,6 +185,8 @@ Several configurations are relevant to microbatch models, and some are required:
 | `batch_size` | String (required)  | The granularity of your batches. Supported values are `hour`, `day`, `month`, and `year`             | N/A     |
 | `lookback`   | Integer (optional) | Process X batches prior to the latest bookmark to capture late-arriving records.                                         | `1`     |
+|`concurrent_batches`| Boolean | Configures whether batches can run concurrently (at the same time) or sequentially (one after the other). When set to `True` and conditions are met, batches run in parallel. |`True` or `False`|
+| `{{this}}` | Macro | When the model definition contains `{{ this }}`, batches can't be ran concurrently | Refer to [Concurrent Execution with the `concurrent_batches` configuration](/docs/build/incremental-microbatch#concurrent-execution-with-the-concurrent_batches-configuration) for more information |
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann committed:
Merge branch 'current' into nfiann-rbip

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Quigley Malcolm <QMalcolm@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Quigley Malcolm <QMalcolm@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Quigley Malcolm <QMalcolm@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Quigley Malcolm <QMalcolm@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Quigley Malcolm <QMalcolm@users.noreply.github.com>

nataliefiann committed:
Merge branch 'current' into nfiann-rbip

mirnawong1 committed:
Update incremental-microbatch.md

committing some changes to the parallel

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,123 @@ For now, dbt assumes that all values supplied are in UTC:
+After checking for 1, 2, and 3 &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relation causes conflict. Otherw ...
+:::
+
+EXPANDABLE IDEA OR H3/H4
```
@nataliefiann to work on this section

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,123 @@ For now, dbt assumes that all values supplied are in UTC:
+- parallel is more performant / faster, but means your logic needs to be independent to the order the batches are executed
+- sequential is slower, but means you can calculate things like cumulative metrics in your microbatch models
+
+<!-- You can override the check for `this` by setting `concurrent_batches` to either `True` or `False`. If set to `False`, the batch will be run sequentially. If set to `True` the batch will be run in parallel (assuming [1], [2], and [3])
```
@nataliefiann checking w quigley

runleonarun replied:
I think this can be removed

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,123 @@ For now, dbt assumes that all values supplied are in UTC:
+  - Snowflake
+  - Databricks
+  - More adapters coming soon!
+- The relation in the data warehouse for the model doesn't exist
```
@nataliefiann confirming that relation is the same as {{ this }}

nataliefiann replied:
Hiya @QMalcolm 

I just wanted to ask about this sentence: "The relation in the data warehouse for the model doesn't exist"
Does the relation refer to {{ this }} ?

Kind Regards
Natalie

QMalcolm replied:
To answer the question, yes and no? 🙃 `{{ this }}` is a python [Relation](https://docs.getdbt.com/reference/dbt-classes#relation) object for the current model. A `Relation` object is essentially just a pointer to where something _should_ be in the data warehouse. The key word though in the previous sentence is "should". That is, having a `Relation` object doesn't necessarily mean that something exists in the data warehouse at the path that the `Relation` points to. So yes when we say "the relation in the data warehouse" in this case, we are talking about `{{ this }}` as the relation. But having `{{ this }}` doesn't guarantee existence in the data warehouse.

Another way of thinking about
> The relation in the data warehouse for the model doesn't exist is

is 
> No model exists in the data warehouse yet for the path specified by the relation

mirnawong1 committed:
Update incremental-microbatch.md

mirnawong1 committed:
Update incremental-microbatch.md

mirnawong1 committed:
Update incremental-microbatch.md

mirnawong1 committed:
Update incremental-microbatch.md

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+
+The microbatch strategy offers the benefit of updating a model in smaller, more manageable batches. 
+
+Parallel batch execution means that multiple batches are processed at the same time, instead of one after the other (sequentially) for faster processing of your microbatch models. 
```
```suggestion
Parallel batch execution means that multiple batches are processed at the same time using the `concurrent_batches` config, instead of one after the other (sequentially) for faster processing of your microbatch models. 
```

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+| 5.   | `concurrent_batches` set to `False` |     -   |   ✅ |
+
+
+- After checking for 1, 2, and 3 in the previous table &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relation ...
```
```suggestion
After checking for 1, 2, and 3 in the previous table &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relation causes conflict. 
```

mirnawong1 committed:
Update website/docs/docs/build/incremental-microbatch.md

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+
+After checking for 1, 2, and 3 in the previous table &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relation c ...
+
+- Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
```
```suggestion
Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
```

mirnawong1 committed:
Update website/docs/docs/build/incremental-microbatch.md

mirnawong1 committed:
Update website/docs/docs/build/incremental-microbatch.md

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -179,12 +179,14 @@ It does not matter whether the table already contains data for that day. Given t
 | `begin`      |  The "beginning of time" for the microbatch model. This is the starting point for any initial or full-refresh builds. For example, a daily-grain microbatch model run on `2024-10-01` with `begin = '2023-10-01` will process 366 batches (it's a leap year!) plus the batch for "today."        | N/A     | Date   | Required |
 | `batch_size` |  The granularity of your batches. Supported values are `hour`, `day`, `month`, and `year`    | N/A     | String  | Required |
 | `lookback`   | Process X batches prior to the latest bookmark to capture late-arriving records.    | `1`     | Integer | Optional |
+|`concurrent_batches`| Configures whether batches can run concurrently (at the same time) or sequentially (one after the other). Can set to `True` or `False`. When set to `True` and conditions are met, batches run in parallel. | `None` | Boolean | Optional |
```
remember to link to concurrent batches page and 'conditions are met' section

nataliefiann replied:
```suggestion
|`concurrent_batches`| An override for whether batches run concurrently (at the same time) or sequentially (one after the other). | `None` | Boolean | Optional |
```

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+
+EXPANDABLE IDEA OR H3/H4
+
+ADD GUIDE/INFORMATION ON WHEN TO USE PARALLEL AND WHEN TO USE SEQUENTIAL FOR USERS. 
```
```suggestion
Parallel batch execution or sequential?
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+
+Parallel batch execution or sequential?
+
+- parallel is more performant / faster, but means your logic needs to be independent to the order the batches are executed
```
```suggestion
- Parallel batch execution is faster but requires logic that is independent of batch execution order. For example, if you are developing a data processing pipeline for a system that processes user transactions in batches, each batch is executed in parallel for better performance. However, the logic used to process each transaction must not depend on the order of how batches are executed or completed.
```

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+Parallel batch execution or sequential?
+
+- parallel is more performant / faster, but means your logic needs to be independent to the order the batches are executed
+- sequential is slower, but means you can calculate things like cumulative metrics in your microbatch models
```
```suggestion
- Sequential processing isn't as performant and is slower however, it enables calculations such as cumulative metrics in microbatch models. Since cumulative metrics require data to be processed in the correct order to ensure each step builds on the previous one, sequential processing is ideal. 
```

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+
+After checking for 1, 2, and 3 in the previous table &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relation c ...
+
+Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
```
Hiya @QMalcolm 

I also wanted to ask about line 317. In the git issue, it advises: 

"If neither (1) nor (2) was hit, then we check if a config concurrent_batches is set for the model. If the value for that config is True then we run the batches in parallel, if False we run the batches sequentially.

If however concurrent_batches is None (i.e. not set), then we check if the model jinja contains a reference to this. If it references this then we run the batches sequentially. Otherwise, we run them in parallel."

Whereas, from your notes, it advises:

"After [1], [2], and [3] we check if the [this jinja function](https://docs.getdbt.com/reference/dbt-jinja-functions/this) is invoked in the model. If this is used, then the batch will be run sequentially, as it may be that your batch depends on the existence of prior batches. If this isn't used, the batch will be run in parallel.

You can override the check for this by setting concurrent_batches to either True or False. If set to False, the batch will be run sequentially. If set to True the batch will be run in parallel (assuming [1], [2], and [3])..."

I just wanted to double check if they meant the same thing as I was a little confused.

Kind Regards
Natalie

QMalcolm replied:
Great question! Happy to clarify ❤️ Also sorry for the confusion 😅 

Those two things are saying the same thing. The first one walks through the actual logical order of operations that happens in core. The second is framed from how the end user should think about it. Basically, the user in ~98% of cases _shouldn't think about or set_ `concurrent_batches`. Core does its best to _automatically detect_ whether a batch can be run in parallel or not. The config `concurrent_batches` is an "escape hatch" to allow the user to say "I actually think these batches should/shouldn't be run in parallel". `concurrent_batches` isn't a _gate_ but instead an _override_.

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+
+EXPANDABLE IDEA OR H3/H4
+
+Parallel batch execution or sequential?
```
```suggestion
### Parallel batch execution or sequential?
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
+
+
+EXPANDABLE IDEA OR H3/H4
```
```suggestion
<Expandable alt_header="Parallel batch execution">
Parallel batch execution is faster but requires logic that is independent of batch execution order. For example, if you are developing a data processing pipeline for a system that processes user transactions in batches, each batch is executed in parallel for better performance. However, the logic used to process each transaction must not depend on the order of how batches are executed or completed.
</Expandable>

<Expandable alt_header="Sequential">
Sequential processing isn't as performant and is slower however, it enables calculations such as cumulative metrics in microbatch models. Since cumulative metrics require data to be processed in the correct order to ensure each step builds on the previous one, sequential processing is ideal. 
</Expandable>
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,132 @@ For now, dbt assumes that all values supplied are in UTC:
+After checking for 1, 2, and 3 in the previous table &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relation c ...
+
+Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
+
```
```suggestion
### Parallel batch execution or sequential processing?
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,132 @@ For now, dbt assumes that all values supplied are in UTC:
+Sequential processing isn't as performant and is slower however, it enables calculations such as cumulative metrics in microbatch models. Since cumulative metrics require data to be processed in the correct order to ensure each step builds on the previous one, sequential processing is ideal. 
+</Expandable>
+
+### Parallel batch execution or sequential?
```
```suggestion
### Parallel batch execution or sequential processing?
```

nataliefiann replied:
```suggestion
### Parallel batch execution or sequential processing?
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,132 @@ For now, dbt assumes that all values supplied are in UTC:
+After checking for 1, 2, and 3 in the previous table &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relation c ...
+
+Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
+### Parallel batch execution or sequential processing?
```
```suggestion
### Parallel or sequential execution?
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,132 @@ For now, dbt assumes that all values supplied are in UTC:
+Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
+### Parallel or sequential execution?
+
+<Expandable alt_header="Parallel batch execution">
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,131 @@ For now, dbt assumes that all values supplied are in UTC:
+Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
+### Parallel or sequential execution?
+
+Parallel batch execution is faster but requires logic that is independent of batch execution order. For example, if you are developing a data processing pipeline for a system that processes user transactions in batches, each batch is executed in parallel for better performance. However, the logic used to process each transaction must not depend on the order of how batches are executed or complete ...
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,130 @@ For now, dbt assumes that all values supplied are in UTC:
+Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
+### Parallel or sequential execution?
+
+</Expandable>
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,129 @@ For now, dbt assumes that all values supplied are in UTC:
+### Parallel or sequential execution?
+
+
+<Expandable alt_header="Sequential">
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,128 @@ For now, dbt assumes that all values supplied are in UTC:
+### Parallel or sequential execution?
+
+
+Sequential processing isn't as performant and is slower however, it enables calculations such as cumulative metrics in microbatch models. Since cumulative metrics require data to be processed in the correct order to ensure each step builds on the previous one, sequential processing is ideal. 
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,127 @@ For now, dbt assumes that all values supplied are in UTC:
+### Parallel or sequential execution?
+
+
+</Expandable>
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+
+
+
+### Parallel batch execution or sequential processing?
```
```suggestion
Choosing between parallel batch execution and sequential processing depends on the specific requirements of your application. 
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -179,12 +179,14 @@ It does not matter whether the table already contains data for that day. Given t
 | `begin`      |  The "beginning of time" for the microbatch model. This is the starting point for any initial or full-refresh builds. For example, a daily-grain microbatch model run on `2024-10-01` with `begin = '2023-10-01` will process 366 batches (it's a leap year!) plus the batch for "today."        | N/A     | Date   | Required |
 | `batch_size` |  The granularity of your batches. Supported values are `hour`, `day`, `month`, and `year`    | N/A     | String  | Required |
 | `lookback`   | Process X batches prior to the latest bookmark to capture late-arriving records.    | `1`     | Integer | Optional |
+|`concurrent_batches`| Configures whether batches can run concurrently (at the same time) or sequentially (one after the other). Can set to `True` or `False`. When set to `True` and conditions are met, batches run in parallel. | `None` | Boolean | Optional |
```
I worry that people reading the first sentence will think they _have_ to use `concurrent_batches` to get batches to run in parallel. Would changing the sentence to something like 
> An override for whether batches run concurrently (at the same time) or sequentially (one after the other).

make sense?

nataliefiann replied:
Hiya @QMalcolm 

This makes sense. I'll change the sentence to your suggestion.

Kind Regards
Natalie

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+  - Snowflake
+  - Databricks
+  - More adapters coming soon!
+- The relation in the data warehouse for the model doesn't exist
```
I think I introduced some confusion somewhere in my prior review. Having
> - The relation in the data warehouse for the model doesn't exist

under a section titled
> To enable parallel execution, you must meet the following conditions:

makes it sound like a batch can only be executed concurrently / in parallel IFF the relation **doesn't** exists in the data warehouse. However, the opposite is actually true. Running a batch in parallel is only possible if the relation (the table) already exists.

Instead of just changing the line though, I'd advocate for removing the line 🤔 Basically the user has no control over this. Additionally the first batch is always run sequentially, thus guaranteeing the relation exists in the data warehouse for all following batches (which may or may not run in parallel in regards to each other after the first batch finishes).

nataliefiann replied:
Thank you for this @QMalcolm - this makes a lot of sense.

mirnawong1 committed:
Merge branch 'current' into nfiann-rbip

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+| 1.   | **Not** the first batch |  ✅         | -            |
+| 2.   | **Not** the last batch  |  ✅         | -            |
+| 3.   | [Adapter supports](#prerequisites) parallel batches | ✅  | -         |
+| 4.   | `concurrent_batches` set to `True`  |  ✅     |   -  |
+| 5.   | `concurrent_batches` set to `False` |     -   |   ✅ |
```
I worry that [4] and [5] here muddy the water and makes it seem like the user _needs_ to take action on setting `concurrent_batches` to either `True` or `False`. I don't think it should be part of this table.

QMalcolm reviewed (commented):
I think I created some confusion in my feedback via github and conversations in slack. We should not be pushing the user to set the `concurrent_batches`config on every microbatch model. We should be framing it as an override available to them if they think core is choosing incorrectly to run things in parallel or not for their use case. That is, the default behavior of microbatch model execution is such that core tries to detect whether a batch can be run in parallel or not. We believe for ~98% of use cases, our detection will be _correct_. If the user thinks we're doing the wrong thing, they can set `microbatch_concurrency` for their desired execution style.

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,126 @@ For now, dbt assumes that all values supplied are in UTC:
+  - Snowflake
+  - Databricks
+  - More adapters coming soon!
+- The relation in the data warehouse for the model doesn't exist
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,125 @@ For now, dbt assumes that all values supplied are in UTC:
+| 1.   | **Not** the first batch |  ✅         | -            |
+| 2.   | **Not** the last batch  |  ✅         | -            |
+| 3.   | [Adapter supports](#prerequisites) parallel batches | ✅  | -         |
+| 4.   | `concurrent_batches` set to `True`  |  ✅     |   -  |
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,124 @@ For now, dbt assumes that all values supplied are in UTC:
+| 1.   | **Not** the first batch |  ✅         | -            |
+| 2.   | **Not** the last batch  |  ✅         | -            |
+| 3.   | [Adapter supports](#prerequisites) parallel batches | ✅  | -         |
+| 5.   | `concurrent_batches` set to `False` |     -   |   ✅ |
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

mirnawong1 committed:
Merge branch 'current' into nfiann-rbip

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,123 @@ For now, dbt assumes that all values supplied are in UTC:
+
+The microbatch strategy offers the benefit of updating a model in smaller, more manageable batches. 
+
+Parallel batch execution means that multiple batches are processed at the same time using the `concurrent_batches` config, instead of one after the other (sequentially) for faster processing of your microbatch models. 
```
@nataliefiann @QMalcolm suggesting the below so it's explicitly clear that dbt auto-detects this most of the time and that the config isn't required
```suggestion
Parallel batch execution means that multiple batches are processed at the same time, instead of one after the other (sequentially) for faster processing of your microbatch models.  

dbt automatically detects whether a batch can be run in parallel in most cases, which means you don’t need to configure this setting. However, the `concurrent_batches` config is available as an override (not a gate), allowing you to specify whether batches should or shouldn’t be run in parallel in specific cases.
```

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,123 @@ For now, dbt assumes that all values supplied are in UTC:
+After checking for 1, 2, and 3 in the previous table &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relation c ...
+
+Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
+### Parallel or sequential execution?
```
do we need a question mark here?
```suggestion
### Parallel or sequential execution
```

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,123 @@ For now, dbt assumes that all values supplied are in UTC:
+
+
+
+Choosing between parallel batch execution and sequential processing depends on the specific requirements of your application. 
```
not sure what we mean about application here
```suggestion
Choosing between parallel batch execution and sequential processing depends on the specific requirements of your use case. 
```

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,123 @@ For now, dbt assumes that all values supplied are in UTC:
+
+Choosing between parallel batch execution and sequential processing depends on the specific requirements of your application. 
+
+- Parallel batch execution is faster but requires logic that is independent of batch execution order. For example, if you are developing a data processing pipeline for a system that processes user transactions in batches, each batch is executed in parallel for better performance. However, the logic used to process each transaction must not depend on the order of how batches are executed or comple ...
```
```suggestion
- Parallel batch execution is faster but requires logic that's independent of batch execution order. For example, if you're developing a data pipeline for a system that processes user transactions in batches, each batch is executed in parallel for better performance. However, the logic used to process each transaction shouldn't depend on the order of how batches are executed or completed.
```

mirnawong1 committed:
Update website/docs/docs/build/incremental-microbatch.md

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,123 @@ For now, dbt assumes that all values supplied are in UTC:
+Choosing between parallel batch execution and sequential processing depends on the specific requirements of your application. 
+
+- Parallel batch execution is faster but requires logic that's independent of batch execution order. For example, if you're developing a data pipeline for a system that processes user transactions in batches, each batch is executed in parallel for better performance. However, the logic used to process each transaction shouldn't depend on the order of how batches are executed or completed.
+- Sequential processing isn't as performant and is slower however, it enables calculations such as cumulative metrics in microbatch models. Since cumulative metrics require data to be processed in the correct order to ensure each step builds on the previous one, sequential processing is ideal. 
```
```suggestion
- Sequential processing is slower but essential for calculations like [cumulative metrics](/docs/build/cumulative)  in microbatch models. It processes data in the correct order, allowing each step to build on the previous one.
```

mirnawong1 committed:
Update website/docs/docs/build/incremental-microbatch.md

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,123 @@ For now, dbt assumes that all values supplied are in UTC:
+
+### Configure `concurrent_batches` 
+
+If you meet all the [conditions](#prerequisites), set the `concurrent_batches` config in your `dbt_project.yml` or incremental microbatch model `.sql` file to run batches in parallel:
```
@nataliefiann @QMalcolm changed this so it's clear that the config is an override option
```suggestion
By default, dbt auto-detects whether batches can run in parallel for microbatch models, and this works correctly in most cases. However, you can override dbt's detection by setting the `concurrent_batches` config in your `dbt_project.yml` or model `.sql` file to specify parallel or sequential execution, given you meet all the [conditions](#prerequisites):
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>

nataliefiann committed:
Merge branch 'current' into nfiann-rbip

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,125 @@ For now, dbt assumes that all values supplied are in UTC:
+
+Choosing between parallel batch execution and sequential processing depends on the specific requirements of your use case. 
+
+- Parallel batch execution is faster but requires logic that's independent of batch execution order. For example, if you're developing a data pipeline for a system that processes user transactions in batches, each batch is executed in parallel for better performance. However, the logic used to process each transaction shouldn't depend on the order of how batches are executed or completed.
```
This is great! 💯 🚀

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -179,12 +179,14 @@ It does not matter whether the table already contains data for that day. Given t
 | `begin`      |  The "beginning of time" for the microbatch model. This is the starting point for any initial or full-refresh builds. For example, a daily-grain microbatch model run on `2024-10-01` with `begin = '2023-10-01` will process 366 batches (it's a leap year!) plus the batch for "today."        | N/A     | Date   | Required |
 | `batch_size` |  The granularity of your batches. Supported values are `hour`, `day`, `month`, and `year`    | N/A     | String  | Required |
 | `lookback`   | Process X batches prior to the latest bookmark to capture late-arriving records.    | `1`     | Integer | Optional |
+|`concurrent_batches`| An override for whether batches run concurrently (at the same time) or sequentially (one after the other). | `None` | Boolean | Optional |
```
adding a link to the page @nataliefiann 
```suggestion
|[`concurrent_batches`](/reference/resource-properties/concurrent_batches)| An override for whether batches run concurrently (at the same time) or sequentially (one after the other). | `None` | Boolean | Optional |
```

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +282,125 @@ For now, dbt assumes that all values supplied are in UTC:
+
+Depending on your use case, configuring your microbatch models to run in parallel offer faster processing, in comparison to running batches sequentially.
+
 ## How `microbatch` compares to other incremental strategies?
```
Can we add to the following to the `## How microbatch compares to other incremental strategies` section:

> As data warehouses roll out new operations for concurrently replacing/upserting data partitions, we may find that the new operation for the data warehouse is more efficient than what the adapter uses for microbatch. In such instances we reserve the right the update the default operation for microbatch, so long as it works as intended/documented for models that fit the microbatch paradigm.

Thank you! ❤️

nataliefiann replied:
Hiya @QMalcolm 

I'll get that added now

Kind Regards
Natalie

mirnawong1 committed:
Update website/docs/docs/build/incremental-microbatch.md

mirnawong1 committed:
Merge branch 'current' into nfiann-rbip

mirnawong1 commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -179,12 +179,15 @@ It does not matter whether the table already contains data for that day. Given t
+| `begin`      |  The "beginning of time" for the microbatch model. This is the starting point for any initial or full-refresh builds. For example, a daily-grain microbatch model run on `2024-10-01` with `begin = '2023-10-01` will process 366 batches (it's a leap year!) plus the batch for "today."        | N/A     | Date   | Required |
+| `batch_size` |  The granularity of your batches. Supported values are `hour`, `day`, `month`, and `year`    | N/A     | String  | Required |
+| `lookback`   | Process X batches prior to the latest bookmark to capture late-arriving records.    | `1`     | Integer | Optional |
+|[`concurrent_batches`](/reference/resource-properties/concurrent_batches)| An override for whether batches run concurrently (at the same time) or sequentially (one after the other). | `None` | Boolean | Optional |
```
```suggestion
| `concurrent_batches` | An override for whether batches run concurrently (at the same time) or sequentially (one after the other). | `None` | Boolean | Optional |
```

mirnawong1 committed:
Update website/docs/docs/build/incremental-microbatch.md

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +283,125 @@ For now, dbt assumes that all values supplied are in UTC:
+- You use the following supported adapters:
+  - Snowflake
+  - Databricks
+  - More adapters coming soon!
```
Can we add a note about `More adapters coming soon!`?

Maybe like
```
- More adapters coming soon!*

*We'll be continuing to test and add concurrency support for adapters. This means that some adapters might get concurrency support _after_ the 1.9 initial release.

nataliefiann replied:
```suggestion
  - More adapters coming soon! 
  
We'll be continuing to test and add concurrency support for adapters. This means that some adapters might get concurrency support _after_ the 1.9 initial release.
```

QMalcolm commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +283,125 @@ For now, dbt assumes that all values supplied are in UTC:
+
+After checking for 1, 2, and 3 in the previous table &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relation c ...
+
+Otherwise, if the `concurrent_batches` value isn't set _and_ `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel.  
```
The current wording might make it sound like batches won't run in parallel if a value for `concurrent_batches` is set. Something like the following might remove the ambiguity 🙂 
```suggestion
Otherwise, if `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel. This can be overriden by setting a value for `concurrent_batches`.
```

QMalcolm reviewed (commented):
This looks soooooo good! ❤️ Thank you so much for all the work you've put into this, and for bearing with me 🙂 Some small additions, but otherwise I think we should be good to go!

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

Co-authored-by: Quigley Malcolm <QMalcolm@users.noreply.github.com>

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +283,127 @@ For now, dbt assumes that all values supplied are in UTC:
+</Tabs>
+
+Depending on your use case, configuring your microbatch models to run in parallel offer faster processing, in comparison to running batches sequentially.
+
```
```suggestion
### How microbatch compares to other incremental strategies

As data warehouses roll out new operations for concurrently replacing/upserting data partitions, we may find that the new operation for the data warehouse is more efficient than what the adapter uses for microbatch. In such instances we reserve the right the update the default operation for microbatch, so long as it works as intended/documented for models that fit the microbatch paradigm.
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +283,129 @@ For now, dbt assumes that all values supplied are in UTC:
+  - Databricks
+  - More adapters coming soon! 
+  
+We'll be continuing to test and add concurrency support for adapters. This means that some adapters might get concurrency support _after_ the 1.9 initial release.
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +283,128 @@ For now, dbt assumes that all values supplied are in UTC:
+- You use the following supported adapters:
+  - Snowflake
+  - Databricks
+  - More adapters coming soon! 
```
```suggestion
  - More adapters coming soon! 
     - We'll be continuing to test and add concurrency support for adapters. This means that some adapters might get concurrency support _after_ the 1.9 initial release.
  
```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann committed:
Added nested bullet

nataliefiann committed:
Merge branch 'nfiann-rbip' of https://github.com/dbt-labs/docs.getdbt.com into nfiann-rbip

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +283,133 @@ For now, dbt assumes that all values supplied are in UTC:
+  - More adapters coming soon!
+    - We'll be continuing to test and add concurrency support for adapters. This means that some adapters might get concurrency support _after_ the 1.9 initial release.
+    
+  - More adapters coming soon! 
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +283,132 @@ For now, dbt assumes that all values supplied are in UTC:
+  - More adapters coming soon!
+    - We'll be continuing to test and add concurrency support for adapters. This means that some adapters might get concurrency support _after_ the 1.9 initial release.
+    
+     - We'll be continuing to test and add concurrency support for adapters. This means that some adapters might get concurrency support _after_ the 1.9 initial release.
```
```suggestion

```

nataliefiann committed:
Update website/docs/docs/build/incremental-microbatch.md

nataliefiann committed:
Merge branch 'current' into nfiann-rbip

nataliefiann committed:
Updated upgrading to v1.9 guide to included parallel batch execution and added links to incremental microbatch page (#6608)

## What are you changing in this pull request and why?

I've created this PR to show updates to upgrading to v1.9 guide to
included parallel batch execution and added links to incremental
microbatch page

## Checklist
- [ ] I have reviewed the [Content style
guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md)
so my content adheres to these guidelines.
- [ ] The topic I'm writing about is for specific dbt version(s) and I
have versioned it according to the [version a whole
page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version)
and/or [version a block of
content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content)
guidelines.
- [ ] I have added checklist item(s) to this list for anything anything
that needs to happen before this PR is merged, such as "needs technical
review" or "change base branch."
- [ ] The content in this PR requires a dbt release note, so I added one
to the [release notes
page](https://docs.getdbt.com/docs/dbt-versions/dbt-cloud-release-notes).
<!--
PRE-RELEASE VERSION OF dbt (if so, uncomment):
- [ ] Add a note to the prerelease version [Migration
Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/docs/dbt-versions/core-upgrade)
-->
<!-- 
ADDING OR REMOVING PAGES (if so, uncomment):
- [ ] Add/remove page in `website/sidebars.js`
- [ ] Provide a unique filename for new pages
- [ ] Add an entry for deleted pages in `website/vercel.json`
- [ ] Run link testing locally with `npm run build` to update the links
that point to deleted pages
-->

<!-- vercel-deployment-preview -->
---
🚀 Deployment available! Here are the direct links to the updated files:


-
https://docs-getdbt-com-git-new-branch-name-1-dbt-labs.vercel.app/docs/build/incremental-microbatch
-
https://docs-getdbt-com-git-new-branch-name-1-dbt-labs.vercel.app/docs/dbt-versions/core-upgrade/06-upgrading-to-v1.9

<!-- end-vercel-deployment-preview -->

---------

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>
Co-authored-by: Leona B. Campbell <3880403+runleonarun@users.noreply.github.com>

nataliefiann committed:
Created concurrent batches page (#6601)

## What are you changing in this pull request and why?

I've created a new page to describe how the concurrent batches config
works

## Checklist
- [ ] I have reviewed the [Content style
guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md)
so my content adheres to these guidelines.
- [ ] The topic I'm writing about is for specific dbt version(s) and I
have versioned it according to the [version a whole
page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version)
and/or [version a block of
content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content)
guidelines.
- [ ] I have added checklist item(s) to this list for anything anything
that needs to happen before this PR is merged, such as "needs technical
review" or "change base branch."
- [ ] The content in this PR requires a dbt release note, so I added one
to the [release notes
page](https://docs.getdbt.com/docs/dbt-versions/dbt-cloud-release-notes).
<!--
PRE-RELEASE VERSION OF dbt (if so, uncomment):
- [ ] Add a note to the prerelease version [Migration
Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/docs/dbt-versions/core-upgrade)
-->
<!-- 
ADDING OR REMOVING PAGES (if so, uncomment):
- [ ] Add/remove page in `website/sidebars.js`
- [ ] Provide a unique filename for new pages
- [ ] Add an entry for deleted pages in `website/vercel.json`
- [ ] Run link testing locally with `npm run build` to update the links
that point to deleted pages
-->

<!-- vercel-deployment-preview -->
---
🚀 Deployment available! Here are the direct links to the updated files:


-
https://docs-getdbt-com-git-new-branch-name-dbt-labs.vercel.app/docs/build/incremental-microbatch
-
https://docs-getdbt-com-git-new-branch-name-dbt-labs.vercel.app/docs/core/connect-data-platform/mssql-setup
-
https://docs-getdbt-com-git-new-branch-name-dbt-labs.vercel.app/docs/dbt-versions/core-upgrade/06-upgrading-to-v1.9
-
https://docs-getdbt-com-git-new-branch-name-dbt-labs.vercel.app/reference/resource-properties/concurrent_batches

<!-- end-vercel-deployment-preview -->

---------

Co-authored-by: Mirna Wong <89008547+mirnawong1@users.noreply.github.com>
Co-authored-by: Leona B. Campbell <3880403+runleonarun@users.noreply.github.com>

runleonarun commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +284,131 @@ For now, dbt assumes that all values supplied are in UTC:
+
+### Prerequisites
+
+To enable parallel execution, you must meet the following conditions:
```
```suggestion
To enable parallel execution, you must:
```

runleonarun commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +284,131 @@ For now, dbt assumes that all values supplied are in UTC:
+
+To enable parallel execution, you must meet the following conditions:
+
+- You use the following supported adapters:
```
```suggestion
- Use a supported adapter:
```

runleonarun commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,6 +284,131 @@ For now, dbt assumes that all values supplied are in UTC:
+    
+  
+  
+- You meet [additional conditions](#how-parallel-batch-execution-works) mentioned in the next section
```
```suggestion
- Meet [additional conditions](#how-parallel-batch-execution-works) described in the next section.
```

runleonarun committed:
Apply suggestions from code review

runleonarun committed:
Update incremental-microbatch.md

runleonarun committed:
Update incremental-microbatch.md

runleonarun commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,7 +284,129 @@ For now, dbt assumes that all values supplied are in UTC:
+
+After checking for the conditions in the previous table &mdash; and if `concurrent_batches` value isn't set, dbt will intelligently auto-detect if the model invokes the [`{{ this }}`](/reference/dbt-jinja-functions/this) Jinja function. If it references `{{ this }}`, the batches will run sequentially since  `{{ this }}` represents the database of the current model and referencing the same relatio ...
+
+Otherwise, if `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel, which can be overriden when you set a value for `concurrent_batches`.
```
```suggestion
Otherwise, if `{{ this }}` isn't detected (and other conditions are met), the batches will run in parallel, which can be overriden when you [set a value for `concurrent_batches`](/reference/resource-properties/concurrent_batches).
```

runleonarun commented on website/docs/docs/build/incremental-microbatch.md:
```diff
@@ -280,7 +284,129 @@ For now, dbt assumes that all values supplied are in UTC:
+
+### Configure `concurrent_batches` 
+
+By default, dbt auto-detects whether batches can run in parallel for microbatch models, and this works correctly in most cases. However, you can override dbt's detection by setting the `concurrent_batches` config in your `dbt_project.yml` or model `.sql` file to specify parallel or sequential execution, given you meet all the [conditions](#prerequisites):
```
```suggestion
By default, dbt auto-detects whether batches can run in parallel for microbatch models, and this works correctly in most cases. However, you can override dbt's detection by setting the [`concurrent_batches` config](/reference/resource-properties/concurrent_batches) in your `dbt_project.yml` or model `.sql` file to specify parallel or sequential execution, given you meet all the [conditions](#prerequisites):
```

runleonarun committed:
Apply suggestions from code review

runleonarun committed:
Update incremental-microbatch.md

runleonarun reviewed (approved):
OK I think this is ready to merge!
